src/tree/updater_gpu_hist.cu

```diff
@@ -1444,23 +1455,101 @@ class GPUHistMakerSpecialised{
                                           column_sampling_seed));
         });
 
-    // Find the cuts.
+    // Initialize Sketches across all batches
+    common::SketchContainer sketch_container;
+    sketch_container.sketches_.resize(info_->num_col_);
+    sketch_container.col_locks_.resize(info_->num_col_);
+#pragma omp parallel for schedule(static)
+    for (int icol = 0; icol < info_->num_col_; ++icol) {
+      sketch_container.sketches_[icol].Init(info_->num_row_, 1.0 / (8 * param_.max_bin));
+      sketch_container.col_locks_[icol].reset(new std::atomic_flag(ATOMIC_FLAG_INIT)); // NOLINT
+    }
+
+    size_t row_stride(0);  // Max row_stride across the entire dataset
+    size_t processed_size(0);
+
+    // Rows allocated to each GPU in a batch
+    std::vector<std::vector<size_t>> shard_allocations;
     monitor_.StartCuda("Quantiles");
-    common::DeviceSketch(batch, *info_, param_, &hmat_, hist_maker_param_.gpu_batch_nrows);
-    n_bins_ = hmat_.row_ptr.back();
+    for (const auto &batch : dmat->GetRowBatches()) {
+      struct RowReset {
```

Comment:
Please avoid declaring new data types inside function and method code, especially if those data types contain methods themselves. It makes understanding the code more difficult.

Reply:
done - i have refactored this area of code.